EE: Publish to Controller

himdel opened the pull request:
Fixes [AAH-889](https://issues.redhat.com/browse/AAH-889)
API https://github.com/ansible/galaxy_ng/pull/941 (controllers; merged) & https://github.com/ansible/galaxy_ng/pull/957 (tags; merged)
Design https://marvelapp.com/prototype/27ee1cd6/screen/81904383

This allows publishing execution environments to Controller.

The target URL Controller-side is `$CONTROLLER_HOST/#/execution_environments/add?image=$OUR_HOST/$image_name:$tag` or `...?image=$OUR_HOST/sha256:$digest` (`OUR_HOST` is without protocol, as in the `podman pull` command).

(Remaining TODO: "Learn more" link - right url)

![20210921191030](https://user-images.githubusercontent.com/289743/134243279-f7e20798-653a-4a24-b864-2161737aca9c.png)

![20210921203045](https://user-images.githubusercontent.com/289743/134243297-a2e05586-5cb0-458d-bd5a-60206b4f8980.png)

![20210921203103](https://user-images.githubusercontent.com/289743/134243305-74fcb87b-661d-4217-bfd0-bca4ef6d43ad.png)

Accessible from:

* [x] EE list - list action
  * preselects last modified tag
  * ![20210906212648](https://user-images.githubusercontent.com/289743/132261429-cf220f16-3d5a-44b0-b61a-9216eb24b012.png)
* [x] EE detail (Detail, Activity, Images) header
  * preselects last modified tag
  * ![20210906214525](https://user-images.githubusercontent.com/289743/132262160-e74cb8d1-cf0e-41a7-9902-c91b26b9d298.png)
* [x] EE Images list - list action
  * preselects last modified tag for chosen digest
  * ![20210902225354](https://user-images.githubusercontent.com/289743/131926532-0342947d-3020-4a43-a7ec-ca77728c0119.png)
* (no links from container manifest)

---

Extra changes:

Also changes the `Edit tags` button to `Manage tags`, to fit the modal title.

![20210921210000](https://user-images.githubusercontent.com/289743/134247292-1a46f79a-4f5c-4cfa-9d17-2c21914e1204.png)

merge `ImagesAPI` into `ExecutionEnvironmentsAPI` - every image-related method except `list` was already there.

`APISearchTypeAhead`: add `toggleIcon` prop, pass to `Select` (to allow showing the tag icon)

`ShaLabel`: add `grey` & `long` bool props 

`getContainersURL`: moved the "give me the server for podman pull" logic to a helper

`EmptyStateNoData`, `EmptyStateCustom` - allow description to be a react node, not ony string

---

Testing:

~~Needs~~ a change in `galaxy_ng/galaxy_ng/app/settings.py` to see any controllers:
```diff
diff --git a/galaxy_ng/app/settings.py b/galaxy_ng/app/settings.py
index 7a84dfb..beeecac 100644
--- a/galaxy_ng/app/settings.py
+++ b/galaxy_ng/app/settings.py
@@ -314,1 +314,1 @@
-CONNECTED_ANSIBLE_CONTROLLERS = []
+CONNECTED_ANSIBLE_CONTROLLERS = ['https://example.com']
```

OR You can add `PULP_CONNECTED_ANSIBLE_CONTROLLERS="['https://example.com']"` to `.compose.env`

sbuenafe-rh commented:
@himdel I would make sure the spacers between Execution Environment form label, Tag, body text and list filter is 24px

himdel commented:
@sbuenafe-rh Thanks, added spacers in those places .. I'm not sure those are right?

![20210922014603](https://user-images.githubusercontent.com/289743/134270520-985f0dac-967d-470f-89db-277bd13f2213.png)

sbuenafe-rh commented:
@himdel 

- For the Tag dropdown select, can we make the max width equal to the max string length in the dropdown options?
- Drop the SHA digest to a new line with a form label, Digest and align tag dropdown with the SHA (mock has been updated)
- Under the title of the modal, the padding should be 24px 
- The padding between the Execution Environments form label and name (mongo) should be 16px

ZitaNemeckova commented on src/components/execution-environment/publish-to-controller-modal.tsx:
```diff
@@ -0,0 +1,384 @@
+    const notListedMessage = (
+      <>
+        {' '}
+        <Trans>
```
Why not use just t`` instead of <Trans> ?

ZitaNemeckova commented on src/components/execution-environment/publish-to-controller-modal.tsx:
```diff
@@ -0,0 +1,384 @@
+          alerts={alerts}
+          closeAlert={(i) => this.closeAlert(i)}
+        ></AlertList>
+        {loading && t`Loading...`}
```
Why not spinner?

ZitaNemeckova commented on src/components/execution-environment/publish-to-controller-modal.tsx:
```diff
@@ -0,0 +1,384 @@
+            <Flex>
+              <FlexItem>
+                <b>
+                  <Trans>Tag</Trans>
```
Why not t``?

ZitaNemeckova reviewed (approved):
LGTM :+1: Some small comments but I'm ok with merging as is :)

himdel commented:
Thanks, added a loading spinner, made l10n more consistent, rebased, and fiixed *some* of the UX issues...

> For the Tag dropdown select, can we make the max width equal to the max string length in the dropdown options?

Done :)

> Drop the SHA digest to a new line with a form label, Digest and align tag dropdown with the SHA (mock has been updated)

Done, except for aligning the two, I can't use the grid layout to do this properly without also having the filter full size, and setting min-width to the larger length breaks with l10n. Will address in a separate PR :).

> Under the title of the modal, the padding should be 24px

This will also have to be addressed separately, but currently the padding is the same in all the modals in the app, should we change this everywhere?

> The padding between the Execution Environments form label and name (mongo) should be 16px

Done, assuming horizontal padding this time :)

> Top pagination on list should align with filter
> Filter search bar doesn't have to span the entire modal width.

Done, both CompoundFilter and Pagination are on the same line now, AppliedFilter on the next one :).

> For the list of controller URLs, make the padding 16px between each. (Attempting to mimic the compact data list component spacing)

Done, added the underlines and a 8px top padding in addition to the default 8px bottom padding :).

> Question: What are the attributes in dropdown filter? Controller Name and ?? If there are more, then they should be shown on the list which means, you should use the compact data list component.

There's only `Controller name` :). Maybe we should adjust CompundFilter to only show the expander when there are more than 1?

--- 

current state:

![20210923194806](https://user-images.githubusercontent.com/289743/134574344-e9e00447-37d4-4b7c-8f55-53a8f1c23ecc.png)

ZitaNemeckova reviewed (approved):
LGTM :+1:

himdel commented:
Merging now, so that we still have time to rebase everything on top of everything else where needed, @sbuenafe-rh let me know if you have any additional concerns please, I'll fix them in the follow-up PR :)

sbuenafe-rh commented:
@himdel Yep, I think you're right to remove the attribute value until it is more than one. Also, nitpicky thing - align the Tag dropdown with the SHA. Thank you!!
